File: inst/d3/scatterplot/dumbbellchart.js
const x1Col        = options.x1;
const x2Col        = options.x2;
const yCol         = options.y;
const x1Label      = options.x1Label || x1Col;
const x2Label      = options.x2Label || x2Col;
const col1         = options.col1         || "steelblue";
const col2         = options.col2         || "crimson";
const lineCol      = options.lineCol      || "#aaaaaa";
const lineWidth    = options.lineWidth    || 2;
const circleRadius = options.circleRadius || 6;
const sortOrder    = options.sort         || "none";
const xtitle       = options.xtitle       || null;
const xtitleSize   = options.xtitleFontSize || 13;
const title        = options.title        || null;
const titleSize    = options.titleFontSize  || 16;
const xFontSize    = options.xFontSize    || 11;
const yFontSize    = options.yFontSize    || 11;
const showLegend   = options.legend !== false;
const legendSize   = options.legendFontSize || 11;
const fontFamily   = options.font    || "Verdana, Geneva, Tahoma, sans-serif";
const bgcol        = options.bgcol   || "white";
const axisCol      = options.axisCol || "black";

svg.selectAll("*").remove();
svg.style("background-color", bgcol);

// Parse and optionally sort rows
let rows = data
  .map(d => ({
    y:  String(d[yCol]),
    x1: +d[x1Col],
    x2: +d[x2Col]
  }))
  .filter(d => isFinite(d.x1) && isFinite(d.x2));

if (sortOrder === "ascending")  rows.sort((a, b) => a.x1 - b.x1);
if (sortOrder === "descending") rows.sort((a, b) => b.x1 - a.x1);

if (!rows.length) {
  svg.append("text")
    .attr("x", width / 2).attr("y", height / 2)
    .attr("text-anchor", "middle")
    .attr("font-family", fontFamily).attr("font-size", 14).attr("fill", axisCol)
    .text("No data available for dumbbell_chart()");
  return;
}

const yDomain = rows.map(d => d.y);

// Legend height reservation
const legendH = showLegend ? legendSize + 16 : 0;

// Margins
const titleOffset  = title  ? titleSize + 14  : 0;
const xtitleOffset = xtitle ? xtitleSize + 10 : 0;

const margin = {
  top:    titleOffset + legendH + 14,
  right:  circleRadius + 20,
  bottom: 28 + xtitleOffset,
  left:   Math.min(width * 0.28, 160)
};

const chartW = width  - margin.left - margin.right;
const chartH = height - margin.top  - margin.bottom;

// Scales
const allValues = rows.flatMap(d => [d.x1, d.x2]);
const xExtent   = d3.extent(allValues);
const xPad      = (xExtent[1] - xExtent[0]) * 0.06 || 1;

const xScale = d3.scaleLinear()
  .domain([xExtent[0] - xPad, xExtent[1] + xPad])
  .range([0, chartW])
  .nice();

const yScale = d3.scaleBand()
  .domain(yDomain)
  .range([0, chartH])
  .padding(0.4);

const g = svg.append("g")
  .attr("transform", `translate(${margin.left},${margin.top})`);

// Subtle horizontal grid lines aligned with each row
rows.forEach(d => {
  g.append("line")
    .attr("x1", 0).attr("x2", chartW)
    .attr("y1", yScale(d.y) + yScale.bandwidth() / 2)
    .attr("y2", yScale(d.y) + yScale.bandwidth() / 2)
    .attr("stroke", "#eeeeee")
    .attr("stroke-width", 1);
});

// Connecting lines
g.selectAll(".dumbbell-line")
  .data(rows)
  .join("line")
  .attr("class", "dumbbell-line")
  .attr("x1", d => xScale(d.x1))
  .attr("x2", d => xScale(d.x2))
  .attr("y1", d => yScale(d.y) + yScale.bandwidth() / 2)
  .attr("y2", d => yScale(d.y) + yScale.bandwidth() / 2)
  .attr("stroke",       lineCol)
  .attr("stroke-width", lineWidth);

// x1 dots (drawn first — sits behind x2)
g.selectAll(".dot-x1")
  .data(rows)
  .join("circle")
  .attr("class", "dot-x1")
  .attr("cx", d => xScale(d.x1))
  .attr("cy", d => yScale(d.y) + yScale.bandwidth() / 2)
  .attr("r",  circleRadius)
  .attr("fill", col1);

// x2 dots
g.selectAll(".dot-x2")
  .data(rows)
  .join("circle")
  .attr("class", "dot-x2")
  .attr("cx", d => xScale(d.x2))
  .attr("cy", d => yScale(d.y) + yScale.bandwidth() / 2)
  .attr("r",  circleRadius)
  .attr("fill", col2);

// X axis
g.append("g")
  .attr("transform", `translate(0,${chartH})`)
  .call(d3.axisBottom(xScale).ticks(6).tickSizeOuter(0))
  .call(ag => ag.select(".domain").attr("stroke", axisCol))
  .call(ag => ag.selectAll(".tick line").attr("stroke", axisCol))
  .call(ag => ag.selectAll(".tick text")
    .attr("font-size",   xFontSize)
    .attr("font-family", fontFamily)
    .attr("fill",        axisCol));

// Y axis
g.append("g")
  .call(d3.axisLeft(yScale).tickSizeOuter(0).tickSize(0))
  .call(ag => ag.select(".domain").remove())
  .call(ag => ag.selectAll(".tick text")
    .attr("font-size",   yFontSize)
    .attr("font-family", fontFamily)
    .attr("fill",        axisCol)
    .attr("dx",          "-8"));

// Legend (top-left, inside chart area)
if (showLegend) {
  const legendY = titleOffset + 6;
  const dot1X   = margin.left;
  const dot2X   = margin.left + circleRadius * 2 + legendSize * 5;

  // x1 legend item
  svg.append("circle")
    .attr("cx", dot1X).attr("cy", legendY)
    .attr("r",  circleRadius * 0.8)
    .attr("fill", col1);

  svg.append("text")
    .attr("x", dot1X + circleRadius + 4)
    .attr("y", legendY)
    .attr("dominant-baseline", "middle")
    .attr("font-size",   legendSize)
    .attr("font-family", fontFamily)
    .attr("fill",        axisCol)
    .text(x1Label);

  // x2 legend item
  svg.append("circle")
    .attr("cx", dot2X).attr("cy", legendY)
    .attr("r",  circleRadius * 0.8)
    .attr("fill", col2);

  svg.append("text")
    .attr("x", dot2X + circleRadius + 4)
    .attr("y", legendY)
    .attr("dominant-baseline", "middle")
    .attr("font-size",   legendSize)
    .attr("font-family", fontFamily)
    .attr("fill",        axisCol)
    .text(x2Label);
}

// Chart title
if (title) {
  svg.append("text")
    .attr("x", margin.left + chartW / 2)
    .attr("y", titleSize)
    .attr("text-anchor",  "middle")
    .attr("font-size",    titleSize)
    .attr("font-family",  fontFamily)
    .attr("font-weight",  "bold")
    .attr("fill",         axisCol)
    .text(title);
}

// X axis title
if (xtitle) {
  svg.append("text")
    .attr("x", margin.left + chartW / 2)
    .attr("y", height - 6)
    .attr("text-anchor",  "middle")
    .attr("font-size",    xtitleSize)
    .attr("font-family",  fontFamily)
    .attr("fill",         axisCol)
    .text(xtitle);
}

File: R/dumbbell_chart.R
#' Create a dumbbell chart.
#'
#' @description
#' A dumbbell chart (also called a connected dot plot) displays two values per
#' category as dots connected by a horizontal line. It is ideal for showing
#' the change or gap between two groups, time points, or conditions across
#' multiple categories — a clean alternative to grouped bar charts favoured
#' by outlets such as the New York Times and The Economist.
#'
#' @param data The data frame containing the variables to consider.
#' @param x1 The name of the column for the first value (left dot).
#' @param x2 The name of the column for the second value (right dot).
#' @param y The name of the categorical column used as row labels.
#' @param x1Label Legend label for the first value. Defaults to the value of
#' \code{x1}.
#' @param x2Label Legend label for the second value. Defaults to the value of
#' \code{x2}.
#' @param col1 Color of the first dot. Defaults to \code{"steelblue"}.
#' @param col2 Color of the second dot. Defaults to \code{"crimson"}.
#' @param lineCol Color of the line connecting the two dots. Defaults to
#' \code{"#aaaaaa"}.
#' @param lineWidth Width of the connecting line in pixels. Defaults to
#' \code{2}.
#' @param circleRadius Radius of the dots in pixels. Defaults to \code{6}.
#' @param sort Whether to sort the rows. One of \code{"none"} (default),
#' \code{"ascending"}, or \code{"descending"}, ordered by the \code{x1}
#' value.
#' @param xtitle Optional. Title of the x-axis.
#' @param xtitleFontSize Font size of the x-axis title. Defaults to \code{13}.
#' @param title Optional. Title of the chart.
#' @param titleFontSize Font size of the chart title. Defaults to \code{16}.
#' @param xFontSize Font size of the x-axis tick labels. Defaults to \code{11}.
#' @param yFontSize Font size of the y-axis category labels. Defaults to
#' \code{11}.
#' @param legend Whether to display a legend. Defaults to \code{TRUE}.
#' @param legendFontSize Font size of the legend text. Defaults to \code{11}.
#' @param font The font family used for all text. Defaults to
#' \code{"Verdana, Geneva, Tahoma, sans-serif"}.
#' @param bgcol Background color of the SVG. Defaults to \code{"white"}.
#' @param axisCol Color of axis lines, ticks, and labels. Defaults to
#' \code{"black"}.
#' @param width Optional. The width of the SVG output.
#' @param height Optional. The height of the SVG output.
#'
#' @return An SVG dumbbell chart.
#' @export
#'
#' @examples
#' # Life expectancy change between 1952 and 2007 for selected countries
#' life_exp <- data.frame(
#'   country = c("Brazil", "China", "Egypt", "India",
#'               "Japan", "Mexico", "Nigeria", "Turkey"),
#'   year_1952 = c(50.9, 44.0, 41.9, 37.4, 63.0, 50.8, 36.3, 43.6),
#'   year_2007 = c(72.4, 72.9, 71.3, 64.7, 82.6, 76.2, 46.9, 71.8)
#' )
#'
#' dumbbell_chart(
#'   data     = life_exp,
#'   x1       = "year_1952",
#'   x2       = "year_2007",
#'   y        = "country",
#'   x1Label  = "1952",
#'   x2Label  = "2007",
#'   title    = "Life expectancy: 1952 vs 2007",
#'   xtitle   = "Life expectancy (years)",
#'   sort     = "ascending"
#' )
#'
#' # Comparing city vs highway fuel economy by car class
#' library(ggplot2)
#' library(dplyr)
#'
#' mpg_summary <- summarise(group_by(mpg, class),
#'   city = mean(cty), highway = mean(hwy))
#'
#' dumbbell_chart(
#'   data    = mpg_summary,
#'   x1      = "city",
#'   x2      = "highway",
#'   y       = "class",
#'   x1Label = "City",
#'   x2Label = "Highway",
#'   col1    = "steelblue",
#'   col2    = "darkorange",
#'   title   = "City vs highway fuel economy by class",
#'   xtitle  = "Miles per gallon",
#'   sort    = "ascending"
#' )
dumbbell_chart <- function(
    data,
    x1,
    x2,
    y,
    x1Label       = NULL,
    x2Label       = NULL,
    col1          = "steelblue",
    col2          = "crimson",
    lineCol       = "#aaaaaa",
    lineWidth     = 2,
    circleRadius  = 6,
    sort          = "none",
    xtitle        = NULL,
    xtitleFontSize = 13,
    title         = NULL,
    titleFontSize  = 16,
    xFontSize     = 11,
    yFontSize     = 11,
    legend        = TRUE,
    legendFontSize = 11,
    font          = "Verdana, Geneva, Tahoma, sans-serif",
    bgcol         = "white",
    axisCol       = "black",
    width         = NULL,
    height        = NULL
) {

  if (is.null(data[[x1]])) stop("Column '", x1, "' not found in data.")
  if (is.null(data[[x2]])) stop("Column '", x2, "' not found in data.")
  if (is.null(data[[y]]))  stop("Column '", y,  "' not found in data.")

  if (!is.numeric(data[[x1]])) stop("Column '", x1, "' must be numeric.")
  if (!is.numeric(data[[x2]])) stop("Column '", x2, "' must be numeric.")

  if (!sort %in% c("none", "ascending", "descending")) {
    stop("'sort' must be one of 'none', 'ascending', or 'descending'.")
  }

  if (grepl(";", font)) {
    stop("please remove the ';' character from your font argument")
  }

  if (is.null(x1Label)) x1Label <- x1
  if (is.null(x2Label)) x2Label <- x2

  r2d3::r2d3(
    data   = data,
    script = system.file("d3/scatterplot/dumbbellchart.js", package = "ddplot"),
    options = list(
      x1            = x1,
      x2            = x2,
      y             = y,
      x1Label       = x1Label,
      x2Label       = x2Label,
      col1          = col1,
      col2          = col2,
      lineCol       = lineCol,
      lineWidth     = lineWidth,
      circleRadius  = circleRadius,
      sort          = sort,
      xtitle        = xtitle,
      xtitleFontSize = xtitleFontSize,
      title         = title,
      titleFontSize  = titleFontSize,
      xFontSize     = xFontSize,
      yFontSize     = yFontSize,
      legend        = legend,
      legendFontSize = legendFontSize,
      font          = font,
      bgcol         = bgcol,
      axisCol       = axisCol
    ),
    width  = width,
    height = height
  )
}
